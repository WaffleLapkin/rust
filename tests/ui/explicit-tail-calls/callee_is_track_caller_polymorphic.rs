//@ build-fail
#![expect(incomplete_features)]
#![feature(explicit_tail_calls)]

fn c<T: Trait>() {
    // `T::f` is not known when checking tail calls,
    // so this has to be checked by the backend.
    become T::f();
}

trait Trait {
    #[track_caller]
    // FIXME(explicit_tail_calls): make error point to the tail call, not callee definition
    fn f() {} //~ error: a function marked with `#[track_caller]` cannot be tail-called
}

impl Trait for () {}

fn main() {
    c::<()>();
}
