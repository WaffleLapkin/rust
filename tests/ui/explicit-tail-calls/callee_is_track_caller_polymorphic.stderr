error: a function marked with `#[track_caller]` cannot be tail-called (llvm)
  --> $DIR/callee_is_track_caller_polymorphic.rs:14:5
   |
LL |     fn f() {}
   |     ^^^^^^

error: aborting due to 1 previous error

